server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,193 @@
+            new HashSet<>(),
+            new HashSet<>()
+        );
+    }
+
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings,
+        NodeDuressTrackers nodeDuressTrackers,
+        Map<String, QueryGroupState> stateMap,
+        Set<QueryGroup> activeQueryGroups,
+        Set<QueryGroup> deletedQueryGroups
+    ) {
+        this.taskCancellationService = taskCancellationService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.nodeDuressTrackers = nodeDuressTrackers;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        activeQueryGroups.forEach(queryGroup -> stateMap.putIfAbsent(queryGroup.get_id(), new QueryGroupState()));
+        this.queryGroupStateMap = stateMap;
+        this.queryGroupStateMap.put(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get(), new QueryGroupState());
+        taskCancellationService.setQueryGroupStateMapAccessor(this::getQueryGroupState);
+        clusterService.addStateApplier(this);
+    }
 
-    public QueryGroupService() {
-        this(new HashMap<>());
+    /**
+     * run at regular interval
+     */
+    protected void doRun() {
+        if (workloadManagementSettings.getWlmMode() == WlmMode.DISABLED) {
+            return;
+        }
+        taskCancellationService.cancelTasks(nodeDuressTrackers::isNodeInDuress, activeQueryGroups, deletedQueryGroups);
+        taskCancellationService.pruneDeletedQueryGroups(deletedQueryGroups);
     }
 
-    public QueryGroupService(Map<String, QueryGroupState> queryGroupStateMap) {
-        this.queryGroupStateMap = queryGroupStateMap;
+    private QueryGroupState getQueryGroupState(final String queryGroupId) {
+        return queryGroupStateMap.getOrDefault(queryGroupId, queryGroupStateMap.get(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get()));
+    }
+
+    /**
+     * {@link AbstractLifecycleComponent} lifecycle method
+     */
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Query Sandbox service", e);
+            }
+        }, this.workloadManagementSettings.getQueryGroupServiceRunInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {}
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        // Retrieve the current and previous cluster states
```

Comment:
Do we need `ClusterStateApplier` here or even `ClusterStateListener` suffices?

Reply:
There is not major changes in both of these Interfaces but `ClusterStateApplier`s are called before `ClusterStateListener`. Do you think ClusterStateListener is more suitable for this use case ?

Reply:
We should use `ClusterStateListener` wherever acceptable. `ClusterStateApplier` is for critical stuff failing which will kick the node out of cluster.

Reply:
Sure, We can change it.